docs/contribution_policy.md

```diff
@@ -0,0 +1,111 @@
+# Contribution policy
+
+*(version v0.1.0 - updated 24 Oct 2023)*
+
+!!! note
+
+    This policy is subject to change, please check back regularly.
+
+## Purpose
+
+The purpose of this contribution policy is to provide guidelines for [adding software](software_layer/adding_software.md) to EESSI.
+
+It informs about what requirements must be met in order for software to be eligible for inclusion
+in the EESSI software layer.
+
+## Requirements
+
+The following requirements must be taken into account when adding software to EESSI.
+
+---
+
+### a) Open source software { #open_source_software }
+
+Only **open source software** can be added to the EESSI repository.
+
+Make sure that you are aware of relevant software license, and that redistribution is allowed.
+
+For more information about a specific license, see the [SPDX license list](https://spdx.org/licenses/).
+
+!!! note
+
+    We intend to automatically verify that this requirement is met,
+    by requiring that the [SPDX license identifier](https://spdx.dev/ids/) is provided for all software.
+
+
+---
+
+### b) Built by the bot { #built_by_bot }
+
+All software included in the EESSI repository *must* be built autonomously by [our bot :robot:](bot.md),
+see also the [semi-automatic software installation procedure](software_layer/adding_software.md).
+
+
+---
+
+### c) Supported by EasyBuild { #supported_by_easybuild }
+
+Currently, we require that all software being added to EESSI is supported by the *EasyBuild release* being used
+to perform the installation.
+
+That is, the easyconfig files used for the installation *must be included in the EasyBuild release*.
+
+We do allow the use of [`--from-pr`](https://docs.easybuild.io/integration-with-github/#github_from_pr) and
+[`--include-easyblocks-from-pr`](https://docs.easybuild.io/integration-with-github/#github_include_easyblocks_from_pr)
+to pull in changes required to make the installation work correctly in the EESSI build environment,
+but only if that is strictly required.
+
+
+---
+
+### d) Compiler toolchain { #compiler_toolchain }
+
+A [compiler toolchain](https://docs.easybuild.io/terminology/#toolchains) that is still supported by the latest
+EasyBuild must be used for building the software.
+
+More information on supported toolchains in EasyBuild is available [here](https://docs.easybuild.io/policies/toolchains).
+
+
+---
+
+### e) CPU targets { #cpu_targets }
+
+The software *should* work on all [CPU targets supported by EESSI](software_layer/cpu_targets.md).
+
+Exceptions to this requirement are allowed if technical problems that can not be resolved with reasonable effort
+prevent the installation of the software for specific CPU targets.
+
+
+---
+
+### f) Versions & toolchains { #versions_toolchains }
+
+Recent software versions and toolchains *should* be preferred,
+although the installation of older software versions and toolchains is allowed if sufficiently motivated.
```

Comment:
I'm not keen on this, we open the door to a lot of pain by looking backwards. In general, I do believe we should decide what toolchains to support on a specific compat and stick to it, and probably take the opportunity once a year to update the compat layer

Reply:
That feels like swinging over to the other side a bit too much (by "locking" a toolchain version to a particular EESSI version, and hence compat layer).

Keeping compat layers updated is gradually going to become quite a bit of effort, and so we may want to install an existing toolchain version in a newer compat layer at some point, since that's likely to be a lot less painful than updating a previous compat layer version.

I guess we could stick to "a toolchain version is only installed on top of a single compat layer version" initially, and then see if that makes sense going forward, and then adjust accordingly if needed.

But should this be part of the contribution policy?

Reply:
I'm on the fence on this one. On the one hand, for a good user / contributor experience, discouraging of forbidding the deployment of old toolchains will _probably_ lead to fewer issues. On the other hand: if someone needs an older toolchain, and if it works, I see little reason not to accept a PR like that. 

I had a chat yesterday with someone from a university HPC support staff who asked 'if my user asks for something from a 2019b toolchain, because he needs the particular software version in that toolchain, can I deploy that through EESSI?'. Even though he fully agreed we should give pushback to users requesting this stuff, in the end he's also realistic and just wants to help the researcher. If it were a local software stack, he would have definitely given it a try. If our policy 'forbids' that, he would be pushed to a local solution. I'm a **bit** concerned if it wouldn't turn HPC support staff away from using EESSI if we are too restrictive on these things.

> I guess we could stick to "a toolchain version is only installed on top of a single compat layer version" initially, and then see if that makes sense going forward, and then adjust accordingly if needed.

What do you mean here? If foss-2022a went into 2023.06, it cannot also go into 2023.12? I wouldn't do that. I think Alan's suggestion was more like: we can decide for 2023.06 to support toolchains 2021a onwards, for 2023.12 we do 2021b onwards, etc. So then there is overlap (2022a would still be on top of both versions of the compat layer), but it prevents deployment of very old toolchains on new compat layers (which probably won't work anyway).

The more I think about it, the more I am _not_ in favour of limiting this initially in the policy. I think people will find it's hard to get older compilers to build, and older MPIs to function properly. In practice, I think they might still be blocked from contributing such toolchains, simply because it's impossible to get them to work (on new compat layers, at least). I wouldn't forbid that, just let the technical limitation be the blocker, not our 'rule'. And, I agree we could still mention this (i.e. old stuff is discouraged, likely to cause issues, and the community probably won't be very motivated to help you out) clearly in the policy.

Another advantage of not limiting it initially is that it allows us to build experience, which we can later use to make an informed decision to forbid it after all (if needed).

Reply:
@casparvl We discussed this aspect during the bot sync meeting earlier today, and I've made some changes to the policy in [86f99c3](https://github.com/EESSI/docs/pull/108/commits/86f99c37cbd38e2dbd1adf2b86e76f17fff529b9) based on that.

There are now separate entries for recent toolchains and recent software versions, where the former mentions that contributors should either use already installed toolchains, or submit a motivated support request to get an additional toolchain installed, which puts the ball in our camp, and leaves the door open for contributors.

There was a consensus during the meeting that this makes sense for the initial version of the policy, and we can revise that later should the need arise.
In practice, it's very likely anyway that is will be EESSI admins who will add toolchains.

We also briefly discussed that we should only install the latest N toolchains in the latest version of EESSI at any time, which allows for minimal overlap across EESSI versions (but that's not part of the policy, that's more of an internal thing we should try and stick to).
In practice this will probably mostly be governed by the compatibility of the glibc version in the compat layer and the GCC version in the toolchains...

Reply:
In https://github.com/EESSI/software-layer/pull/371 I've included the ability for both the site _and_ the user to extend EESSI as they see fit. So even if _we_ don't support it (most likely because of the architecture situation) doesn't mean that they can't do that locally if it "just works".